feat: Set defaults for empty basic secrets (#347)

pratikjagrut opened the pull request:
If no file or data is passed while creating a basic secret then default username=username and password=password will be set.
This PR addresses #347 issue.
### Checklist
- [x] The title of this PR would make a good line in Acorn's Release Note's Changelog
- [x] The title of this PR ends with a link to the main issue being address in parentheses, like: `This is a title (#1216)`. [Here's an example](https://github.com/acorn-io/acorn/pull/1199)
- [x] All relevant issues are referenced in the PR description. *NOTE: don't use [GitHub keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) that auto-close issues*
- [x] Commits follow [contributing guidance](https://github.com/acorn-io/acorn/blob/main/CONTRIBUTING.md#commits)
- [x] Automated tests added to cover the changes. If tests couldn't be added, an explanation is provided in the Verification and Testing section
- [ ] Changes to user-facing functionality, API, CLI, and upgrade impacts are clearly called out in PR description
- [ ] PR has at least two approvals before merging (or a reasonable exception, like it's just a docs change)

pratikjagrut commented:
Hey @cjellick, could you review this PR? It's related to your issue. Thanks!

cjellick commented:
My hope/expectation in that issue was that random values would be generated for username and password, as they are when you create a basic secret similarly in an acornfile, see: https://docs.acorn.io/authoring/secrets#basic-secrets

pratikjagrut commented:
@cjellick Thank you for the feedback.

I have incorporated your suggested changes in the updated PR. Additionally, I noticed the comment in the code section you mentioned [here](https://github.com/acorn-io/runtime/blob/main/pkg/secrets/secret.go#L255), which appears to be related to generating a random username and secret from the Acornfile, as you mentioned. In response, I have implemented a new secret generation function based on that comment. Please review the changes and let me know if they meet your approval.

g-linville commented:
@pratikjagrut thank you for your interest in contributing to Acorn! These changes work well for generating default secret values in the CLI, but the issue also includes this:

> So, I also want kubectl create -f secret.yaml to work where secret.yaml is
> ```
> apiVersion: api.acorn.io/v1
> kind: Secret
> metadata:
>   name: staging-db-user-credentials2
>   namespace: acorn
> type: basic
> ```

In order for this to work, there needs to be some logic in the api-server to generate the values there.

The api-server code is in the folder `pkg/server`. In particular we are interested in `pkg/server/registry/apigroups/acorn/secrets`. Here you will see a file called `storage.go` which sets up the strategy for storing Acorn secrets. We use a translation strategy, meaning, that they will be stored as CRDs internally in Kubernetes, but will go through some translation logic before being presented to the user.

It is possible to override the secret creation logic with some custom code to generate these values. If you take a look at the folder `pkg/server/registry/apigroups/acorn/builds`, you can see in its `storage.go` file that it is also using a translation strategy, but it creates another strategy to wrap around the translation strategy specifically for creation. You can see the code for that wrapper in the `strategy.go` file's `Create` function. You can use a pattern like this to generate the default values in the api-server.

Thanks again for your interest in contributing and let me know if you have any questions!

g-linville commented on pkg/secrets/secret.go:
```diff
@@ -252,15 +251,15 @@ func generateBasic(req router.Request, appInstance *v1.AppInstance, secretName s
 		if len(secret.Data[key]) == 0 {
-			// TODO: Improve with more characters (special, upper/lowercase, etc)
-			v, err := randomtoken.Generate()
+			v, err := GenerateRandomSecret(16)
```
nit: the function that this replaces generated a string of length 54, so let's do that here too

```suggestion
			v, err := GenerateRandomSecret(54)
```

pratikjagrut replied:
Even if I make the suggested changes, they won't have any impact since the subsequent code truncates the length to 8 characters for the username and 16 characters for the password.

g-linville commented on pkg/server/registry/apigroups/acorn/secrets/defaultCreate.go:
```diff
@@ -0,0 +1,43 @@
+			return nil, err
+		}
+
+		if secret.Data == nil {
+			secret.Data = map[string][]byte{}
+		}
```
nit: this was already checked in the outer if statement

```suggestion
		secret.Data = map[string][]byte{}
```

pratikjagrut replied:
Good catch. Thanks. :)

g-linville commented on pkg/server/registry/apigroups/acorn/secrets/defaultCreate.go:
nit: I would like this file to be renamed to `strategy.go`, since that is the convention used with other API types.

pratikjagrut replied:
okay.

g-linville reviewed (changes requested):
Looks good, @pratikjagrut! I left a few small comments. Also, I would like for the username/password generation in the CLI code to be removed so that the apiserver can always do it instead.

g-linville commented on pkg/server/registry/apigroups/acorn/secrets/strategy .go:
```diff
@@ -0,0 +1,41 @@
+			return nil, err
+		}
+
+		secret.Data = map[string][]byte{}
+
+		secret.Data["username"] = []byte(username)
+		secret.Data["password"] = []byte(password)
```
nit: one small cleanup

```suggestion
		secret.Data = map[string][]byte{
			"username": []byte(username),
			"password": []byte(password),
		}
```

g-linville replied:
@pratikjagrut I would still like for this to be changed, if you are willing.

g-linville commented on pkg/server/registry/apigroups/acorn/secrets/strategy .go:
nit: the name of this file has a space before the extension

g-linville reviewed (commented):
@pratikjagrut couple more small things, but looks good! I'll have one of the other Acorn Labs guys look over it too.

tylerslaton reviewed (approved):
Looks great! Thank you for the contribution @pratikjagrut!
